### ISSUE TITLE
您的方法的适应范围是：必须存在已配准的图像对吗？

### ISSUE THREAD

**Issue by lcx2002-zju** [Junchen-Yu/SSHNet#2]
您好：
我按照您的建议修改了dataloader,但是还是不能复现出代码，甚至我还试了同模态的图像对在增强时用192*192大小的template_original里的图像，但无监督图像对用固定的template里的图像，依然无法训练出正常的结果。
import numpy as np
from torch.utils.data import Dataset, DataLoader
from glob import glob
import os
import json
import torch
import os.path as osp
import cv2
import albumentations as A
from utils.homo_utils import generate_homo
patch_size = 128
class homo_dataset(Dataset):
    def __init__(self, split, dataset, args):
        self.dataset = dataset
        self.args = args
        self.homo_parameter = {"marginal":32, "perturb":32, "patch_size":128}
        self.transform = A.Compose([A.RandomBrightnessContrast(brightness_limit=0.4, contrast_limit=0, brightness_by_max=False, always_apply=True),
                                    A.Sharpen(always_apply=True)])
        self.augmentation = A.ReplayCompose([A.VerticalFlip(always_apply = False,p = 0.5),
                                             A.HorizontalFlip(always_apply = False,p = 0.5),
                                             A.Rotate(limit=90, interpolation=1, border_mode=4, value=None, mask_value=None, always_apply=False, p=0.5)])
        
        if dataset == 'VIS_IR':
            if split == 'train':   
                root_img2 = '/hhd/wxc data/VIS_IR_drone/train_small_size/VIS'
                root_img1 = '/hhd/wxc data/VIS_IR_drone/train_small_size/IR'
                root_label = '/hhd/wxc data/VIS_IR_drone/train_small_size/label'
            else:
                root_img2 = '/hhd/wxc data/VIS_IR_drone/test_small_size/VIS'
                root_img1 = '/hhd/wxc data/VIS_IR_drone/test_small_size/IR'
                root_label = '/hhd/wxc data/VIS_IR_drone/test_small_size/label'
        if dataset =='VIS_NIR':
            if split == 'train':   
                root_img2 = '/hhd/wxc data/VIS_NIR/train_small_size/VIS'
                root_img1 = '/hhd/wxc data/VIS_NIR/train_small_size/NIR'
                root_label = '/hhd/wxc data/VIS_NIR/train_small_size/label'
            else:
                root_img2 = '/hhd/wxc data/VIS_NIR/test_small_size/VIS'
                root_img1 = '/hhd/wxc data/VIS_NIR/test_small_size/NIR'
                root_label = '/hhd/wxc data/VIS_NIR/test_small_size/label'
        if dataset == 'GE':
            if split == 'train':   
                root_img2 = '/hhd/wxc data/Dataset/Dataset/GoogleEarth/train2014_input'
                root_img1 = '/hhd/wxc data/Dataset/Dataset/GoogleEarth/train2014_template'
                root_label = '/hhd/wxc data/Dataset/Dataset/GoogleEarth/train2014_label'
            else:
                root_img2 = '/hhd/wxc data/Dataset/Dataset/GoogleEarth/val2014_input'
                root_img1 = '/hhd/wxc data/Dataset/Dataset/GoogleEarth/val2014_template'
                root_label = '/hhd/wxc data/Dataset/Dataset/GoogleEarth/val2014_label'
        if dataset == 'GM':
            if split == 'train':   
                root_img2 = '/hhd/wxc data/Dataset/Dataset/GoogleMap/train2014_input'
                root_img1 = '/hhd/wxc data/Dataset/Dataset/GoogleMap/train2014_template'
                root_label = '/hhd/wxc data/Dataset/Dataset/GoogleMap/train2014_label'
            else:
                root_img2 = '/hhd/wxc data/Dataset/Dataset/GoogleMap/val2014_input'
                root_img1 = '/hhd/wxc data/Dataset/Dataset/GoogleMap/val2014_template'
                root_label = '/hhd/wxc data/Dataset/Dataset/GoogleMap/val2014_label'
        if dataset == 'VEDAI':
            if split == 'train':   
                root_img2 = '/hhd/wxc data/VEDAI/train_input'
                root_img1 = '/hhd/wxc data/VEDAI/train_template'
                root_label = '/hhd/wxc data/VEDAI/train_label'
            else:
                root_img2 = '/hhd/wxc data/VEDAI/test_input'
                root_img1 = '/hhd/wxc data/VEDAI/test_template'
                root_label = '/hhd/wxc data/VEDAI/test_label'
        self.dataset = dataset
        if dataset == 'VEDAI':
            self.image_list_img1 = sorted(glob(osp.join(root_img1, '*.png')))
            self.image_list_img2 = sorted(glob(osp.join(root_img2, '*.png')))
            self.image_list_label = sorted(glob(osp.join(root_label, '*.txt')))
        else:
            self.image_list_img1 = sorted(glob(osp.join(root_img1, '*.jpg')))
            self.image_list_img2 = sorted(glob(osp.join(root_img2, '*.jpg')))
            self.image_list_label = sorted(glob(osp.join(root_label, '*.txt')))

    def __len__(self):
        return int(len(self.image_list_img1))

    def __getitem__(self, index):
        img1 = cv2.imread(self.image_list_img1[index])
        img2 = cv2.imread(self.image_list_img2[index])
        img1 = cv2.cvtColor(img1, cv2.COLOR_BGR2RGB)
        img2 = cv2.cvtColor(img2, cv2.COLOR_BGR2RGB)
        with open(self.image_list_label[index], 'r') as outfile:
                data = json.load(outfile)
        u_list = [data['location'][0]['top_left_u'], data['location'][1]['top_right_u'], data['location'][3]['bottom_right_u'], data['location'][2]['bottom_left_u']]
        v_list = [data['location'][0]['top_left_v'], data['location'][1]['top_right_v'], data['location'][3]['bottom_right_v'], data['location'][2]['bottom_left_v']]
        
        points2 = np.stack([np.array(u_list), np.array(v_list)], axis=-1).astype(np.float32)
        
        points_fixed = np.float64([[0,0],[patch_size-1,0],[0,patch_size-1],[patch_size-1,patch_size-1]])
        points1 = np.float64([[32,32],[32+patch_size-1,32],[32,32+patch_size-1],[32+patch_size-1,32+patch_size-1]])#source
    
        disp = np.zeros((4,2))
        disp[0] = points2[0]-points1[0]
        disp[1] = points2[1]-points1[1]
        disp[2] = points2[3]-points1[2]
        disp[3] = points2[2]-points1[3]
        points2  = points1 + disp

        self.homo_parameter["height"], self.homo_parameter["width"], _ = img2.shape
        
        img1_pad = cv2.copyMakeBorder(img1, (192 - 128) // 2, (192 - 128) // 2, (192 - 128) // 2, (192 - 128) // 2, cv2.BORDER_REFLECT_101)
        # print(img1_pad.shape)
        if self.args.randomize:
            patch_img1_warp, patch_img1, gt11, _, _, _, _ = generate_homo(img1_pad, img1_pad, homo_parameter=self.homo_parameter, transform=self.transform)
            patch_img2_warp, patch_img2, gt22, _, _, _, _ = generate_homo(img2, img2, homo_parameter=self.homo_parameter, transform=self.transform)
        else:
            patch_img1_warp, patch_img1, gt11, _, _, _, _ = generate_homo(img1_pad, img1_pad, homo_parameter=self.homo_parameter, transform=None)
            patch_img2_warp, patch_img2, gt22, _, _, _, _ = generate_homo(img2, img2, homo_parameter=self.homo_parameter, transform=None)
        
        patch_img1_eval = img1[:,:]
        patch_img2_eval = img2[32:160,32:160]
        
        patch_img1_eval = torch.from_numpy(patch_img1_eval).float().permute(2, 0, 1)
        patch_img1_eval = patch_img1_eval/255.0

        patch_img2_eval = torch.from_numpy(patch_img2_eval).float().permute(2, 0, 1)
        patch_img2_eval = patch_img2_eval/255.0
        
        gt12_eval = disp.astype(np.float32)
        org_pts_eval = points1.astype(np.float32)
        dst_pts_eval = points2.astype(np.float32)
        large_img1_eval = torch.from_numpy(img1_pad).float().permute(2, 0, 1)/255.0
        large_img2_eval = torch.from_numpy(img2).float().permute(2, 0, 1)/255.0
        
        return {"patch_img1_warp":patch_img1_warp, "patch_img1":patch_img1, "gt11":gt11, 
                "patch_img2_warp":patch_img2_warp, "patch_img2":patch_img2, "gt22":gt22, 
                "patch_img1_eval":patch_img1_eval, "patch_img2_eval":patch_img2_eval, "gt12_eval":gt12_eval,
                "org_pts_eval":org_pts_eval, "dst_pts_eval":dst_pts_eval,
                "large_img1_eval":large_img1_eval, "large_img2_eval":large_img2_eval}

def fetch_dataloader(args, split='train'):
    if split == 'train':
        train_dataset = homo_dataset(split='train', dataset=args.dataset, args=args)
        train_loader = DataLoader(train_dataset, batch_size=args.batch_size, pin_memory=True, shuffle=True, num_workers=16, drop_last=False)
        print('Training with %d image pairs' % len(train_dataset))
    else: 
        train_dataset = homo_dataset(split='test', dataset=args.dataset, args=args)
        train_loader = DataLoader(train_dataset, batch_size=args.batch_size, pin_memory=True, shuffle=True, num_workers=16, drop_last=False)
        print('Test with %d image pairs' % len(train_dataset))
    
    return train_loader

**Comment by Junchen-Yu** [comment]
不需要存在已配准的图像对，请检查修改代码后，log中生成的仿真训练数据是否合理
